haskell-api.md

```diff
@@ -0,0 +1,236 @@
+# Haskell API Conventions
+
+## Stateless
+
+Prefer stateless endpoints. Sessions are used to authorize data access, but not
+choose what data to return.
+
+- Good:
+  - `/3/students?schools.id=x,y,z`
+  - `ApiClient.fetch("/3/students", { "schools.id": me.schoolIds })`
+- Bad:
+  - `/3/school-admins/me/students`
+  - `ApiClient.fetch("/3/school-admins/me/students")`
+
+## Parameters
+
+We want to :heart: query parameters as a way to keep the number of distinct
+endpoints lower by supporting more complex filtering combinations within each
+endpoints.
+
+- Good: `/3/students?schools.id=x,y,z&teachers.id=a,b,c`
+- Bad:
+  - `/3/schools/:id/students`
+  - `/3/teachers/:id/students`
+  - `/3/schools/:id/teachers/:id/students`
+  - etc, forever
+
+### Naming
+
+Filters should match the dotted-path of the filtered attribute in the response.
+
+For example:
+
+```
+/3/teachers
+[ { id:
+  , ...
+  , school:
+    { id:
+    , ...
+    }
+  }
+]
+```
+
+Here, a School filter would be `school.id`: `/3/teachers?school.id=1,2,3`.
+
+```
+/3/students
+[ { id:
+  , ...
+  , schools:
+    [ { id:
+      , ...
+      }
+    ]
+  }
+]
+```
+
+But here, it would be `schools.id`: `/3/students?schools.id=1,2,3`.
+
+If a filter exists that is not for an attribute present in the response, the
+name can be inferred by what it would look like if it were.
+
+### Semantics
+
+Prefer `IN` equality semantics.
+
+- Good: `{attribute}={value1},{value2},{value3},...`
+- Bad: `{attribute}={value}`
+
+This is because:
+
+- It's just as easy to support as direct equality
+- It reduces to the expected `EQUALS` semantics when given one value
+- We think we can get by with only this level "smart" filtering for a while and
+  defer more complexity in this area (e.g. "not", etc).
+
+---
+
+**Aside about URL length limits**: We don't know at what size this will be a
+problem. Standards-wise, there is no limit, but browsers and servers do have
+limits and it's recommended to use 2000 as a reference point [ref].
+
+[ref]: https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers
+
+However, I will say this:
```

Comment:
This seems like a commit message and not something that should be in the guide proper. The guide should tell you what to do, quickly and tersely. This extra exposition just clutters things up.

Reply:
Yup: https://github.com/freckle/guides/pull/39#discussion_r310117292